### PR TITLE
fix: Long filename handling in Preview

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details

- Long filename was overflowing outside the container. Added a component that adds ellipses (...) in the middle of the string. 
- Updated `subtitle` prop to `Header` so that a custom component or string can be passed to `Header`

### Fixed Issues

$ https://github.com/Expensify/App/issues/6873

### Tests
1. Tested with long filenames on Desktop, Web, and Mobile Apps
2. Tested Fluid width to ensure ellipses show only if there isn't space for the whole name to show up.
3. Tested `ReportingTypingIndicator` as the component was updated.

- [x] Verify that no errors appear in the JS console

### QA Steps
1. Launch the app and open any chat
2. Upload an attachment with a very long name
2. Open the attachment with the preview pane
3. Ensure that the filename should be visible with `...` and ensure that the extension should show up.

- [] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

web-file-preview-long-text

#### Mobile Web

mweb-file-preview-long-text

#### Desktop

desktop-file-preview-long-text

#### iOS

ios-file-preview-long-text

#### Android

android-file-preview-long-text

### CODE REVIEW

File: src/components/TextWithEllipses/index.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                    {props.trailingText}
+                </Text>
+            </View>
+        </View>
+    );
+};
+
+TextWithEllipses.propTypes = propTypes;
+TextWithEllipses.defaultProps = defaultProps;
+TextWithEllipses.displayName = 'TextWithEllipses';
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:58]:
```suggestion
TextWithEllipsis.displayName = 'TextWithEllipsis';
```

Similarly update in all places.

File: src/components/TextWithEllipses/index.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    trailingText: PropTypes.string.isRequired,
+
+    /** styling for leading and trailing text */
+    textStyle: stylePropTypes,
+
+    /** styling for leading text View */
+    leadingTextParentStyle: stylePropTypes,
+
+    /** styling for trailing text View */
+    trailingTextParentStyle: stylePropTypes,
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:23]: Unused prop. 

```suggestion
```

File: src/components/TextWithEllipses/index.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        <View style={[styles.flexRow, ...wrapperStyles]}>
+            <View style={[styles.flexShrink1, ...leadingTextParentStyles]}>
+                <Text
+                    style={props.textStyle}
+                    numberOfLines={1}
+                >
+                    {props.leadingText}
+                </Text>
+            </View>
+            <View style={[styles.flexShrink0, ...trailingTextParentStyles]}>
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:47]:
```suggestion
            <View style={styles.flexShrink0}>
```

File: src/components/TextWithEllipses/index.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    trailingTextParentStyle: stylePropTypes,
+
+    /** styling for parent View */
+    wrapperStyle: stylePropTypes,
+};
+
+const defaultProps = {
+    textStyle: {},
+    leadingTextParentStyle: {},
+    trailingTextParentStyle: {},
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:32]:
```suggestion

```

File: src/components/TextWithEllipses/index.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    wrapperStyle: {},
+};
+
+const TextWithEllipses = (props) => {
+    const wrapperStyles = _.isArray(props.wrapperStyle) ? props.wrapperStyle : [props.wrapperStyle];
+    const leadingTextParentStyles = _.isArray(props.leadingTextParentStyle) ? props.leadingTextParentStyle : [props.leadingTextParentStyle];
+    const trailingTextParentStyles = _.isArray(props.trailingTextParentStyle) ? props.trailingTextParentStyle : [props.trailingTextParentStyle];
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:39]:
```suggestion

```

Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:39]: Why don't we create a small util for this and then use to directly in props. 

Something like 
```
style={[...parseStyleAsArray(props.wrapperStyle])}

```

File: src/components/TextWithEllipses/index.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    /** styling for leading and trailing text */
+    textStyle: stylePropTypes,
+
+    /** styling for leading text View */
+    leadingTextParentStyle: stylePropTypes,
+
+    /** styling for trailing text View */
+    trailingTextParentStyle: stylePropTypes,
+
+    /** styling for parent View */
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:25]:
```suggestion
    /** Styles for parent View */
```

File: src/components/TextWithEllipses/index.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    /** leading text, first text of the group */
+    leadingText: PropTypes.string.isRequired,
+
+    /** trailing text, second text of the group */
+    trailingText: PropTypes.string.isRequired,
+
+    /** styling for leading and trailing text */
+    textStyle: stylePropTypes,
+
+    /** styling for leading text View */
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:19]:
```suggestion
    /** Styles for leading text View */
```

File: src/components/TextWithEllipses/index.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import stylePropTypes from '../../styles/stylePropTypes';
+
+const propTypes = {
+    /** leading text, first text of the group */
+    leadingText: PropTypes.string.isRequired,
+
+    /** trailing text, second text of the group */
+    trailingText: PropTypes.string.isRequired,
+
+    /** styling for leading and trailing text */
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:16]:
```suggestion
    /** Styles for leading and trailing text */
```

File: src/components/TextWithEllipses/index.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import _ from 'underscore';
+import Text from '../Text';
+import styles from '../../styles/styles';
+import stylePropTypes from '../../styles/stylePropTypes';
+
+const propTypes = {
+    /** leading text, first text of the group */
+    leadingText: PropTypes.string.isRequired,
+
+    /** trailing text, second text of the group */
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:13]:
```suggestion
    /** Text after ellipsis */
```

File: src/components/TextWithEllipses/index.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {View} from 'react-native';
+import _ from 'underscore';
+import Text from '../Text';
+import styles from '../../styles/styles';
+import stylePropTypes from '../../styles/stylePropTypes';
+
+const propTypes = {
+    /** leading text, first text of the group */
```
Comment (parasharrajat) [inline src/components/TextWithEllipses/index.js:10]:
```suggestion
    /** Leading text before the ellipsis */
```

File: src/components/AttachmentModal.js [hunk 11]
```diff
@@ -73,6 +74,21 @@ class AttachmentModal extends PureComponent {
...
         this.closeConfirmModal = this.closeConfirmModal.bind(this);
         this.isValidSize = this.isValidSize.bind(this);
+        this.getSplitFileName = this.getSplitFileName.bind(this);
+    }
+
+    /**
+     * Returns the filename split into leading and trailing text
+     * @returns {Object}
+     */
+    getSplitFileName() {
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js:84]:
```suggestion
    splitExtensionFromFileName() {
```

File: src/components/AttachmentModal.js [hunk 12]
```diff
@@ -73,6 +74,21 @@ class AttachmentModal extends PureComponent {
...
+        const fileName = this.props.originalFileName ? this.props.originalFileName : lodashGet(this.state, 'file.name', '');
+        const splittedFileName = fileName.split('.');
+        const trailingText = splittedFileName.pop();
+        const leadingText = splittedFileName.join('.');
+        return {
+            leadingText, trailingText,
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js]:
```suggestion
            fileName, fileExtension,
```

Lets edit the above variables same as these names.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/Header.js [hunk 13]
```diff
@@ -22,12 +23,12 @@ const defaultProps = {
 };
 const Header = props => (
     <View style={[styles.flex1, styles.flexRow]}>
-        <View>
+        <View style={styles.w100}>
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/components/Header.js:26]: If I don't set this 100%, the View overflows the width.

image

[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/Header.js:26]: lets make it flex1

File: src/components/Header.js [hunk 14]
```diff
@@ -10,7 +11,7 @@ const propTypes = {
     title: PropTypes.string.isRequired,
 
     /** Subtitle of the header */
-    subtitle: PropTypes.string,
+    subtitle: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/components/Header.js:14]: I've used `PropTypes.element` now. Does that work?

File: src/components/Header.js [hunk 15]
```diff
@@ -22,12 +23,12 @@ const defaultProps = {
...
 const Header = props => (
     <View style={[styles.flex1, styles.flexRow]}>
-        <View>
+        <View style={styles.w100}>
             <Text numberOfLines={2} style={[styles.headerText, styles.textLarge]}>
                 {props.title}
             </Text>
             {/* If there's no subtitle then display a fragment to avoid an empty space which moves the main title */}
-            {props.subtitle ? <Text style={[styles.mutedTextLabel]}>{props.subtitle}</Text> : <></> }
+            {_.isString(props.subtitle) ? <Text style={[styles.mutedTextLabel]}>{props.subtitle}</Text> : props.subtitle }
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/components/Header.js:31]: I'll update, but any specific reason to split? Especially when the earlier one was also a 1-liner?

File: src/components/TextWithEllipses/index.js (hunk 5 above)
Comment (mananjadhav) [reply src/components/TextWithEllipses/index.js]: I've added a function to `StyleUtils`. Does that work?

File: src/components/Header.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/components/Header.js:26]: Make what `flex-row`? You mean `<View style={styles.w100}>` ==> `<View style={styles.flexRow}>`

**Comment by mananjadhav** [comment]
@parasharrajat PR updated with 1 pending clarification and 1-2 questions.

File: src/components/Header.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/Header.js]: For better visibility, following styling guide.

File: src/components/Header.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/Header.js:26]: Yeah..

File: src/components/Header.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/Header.js]: Yup. Thanks.

File: src/components/TextWithEllipses/index.js (hunk 5 above)
Comment (parasharrajat) [reply src/components/TextWithEllipses/index.js]: Yeah.

File: src/components/AttachmentModal.js [hunk 16]
```diff
@@ -89,6 +91,19 @@ class AttachmentModal extends PureComponent {
         return modalType;
     }
 
+    /**
+     * Returns the filename split into leading and trailing text
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js:95]: This needs to be adjusted as well.

File: src/components/AttachmentModal.js [hunk 17]
```diff
@@ -89,6 +91,19 @@ class AttachmentModal extends PureComponent {
...
+     * @returns {Object}
+     */
+    splitExtensionFromFileName() {
+        const fullFileName = this.props.originalFileName ? this.props.originalFileName : lodashGet(this.state, 'file.name', '');
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js:99]: Lets trim the string here.

File: src/components/AttachmentModal.js [hunk 18]
```diff
@@ -148,7 +165,14 @@ class AttachmentModal extends PureComponent {
                         shouldShowDownloadButton={!this.props.isUploadingAttachment}
                         onDownloadButtonPress={() => fileDownload(sourceURL)}
                         onCloseButtonPress={() => this.setState({isModalOpen: false})}
-                        subtitle={this.props.originalFileName ? this.props.originalFileName : lodashGet(this.state, 'file.name', '')}
+                        subtitle={(
+                            <TextWithEllipsis
+                                leadingText={fileName.trim()}
+                                trailingText={fileExtension ? `.${fileExtension.trim()}` : ''}
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js:171]:
```suggestion
                                trailingText={fileExtension ? `.${fileExtension}` : ''}
```

we don't need to trim here if we do this above.

File: src/components/AttachmentModal.js [hunk 19]
```diff
@@ -89,6 +91,19 @@ class AttachmentModal extends PureComponent {
...
+     */
+    splitExtensionFromFileName() {
+        const fullFileName = this.props.originalFileName ? this.props.originalFileName : lodashGet(this.state, 'file.name', '');
+        const splittedFileName = fullFileName.split('.');
+        const fileExtension = splittedFileName.pop();
+        const fileName = splittedFileName.join('.');
+        return {fileName, fileExtension};
+    }
+
+
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js:106]:
```suggestion
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/Header.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/components/Header.js:26]: That won't work as the header `Text` and the subtitile of will come on next to each other. I can make it `flex1` or `mw100`?

[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/Header.js:26]: looks good to me. Just we should not occupy 100% until content is that long.

[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/components/Header.js]: Got it. I've updated it to `mw100`

**Comment by mananjadhav** [comment]
PR updated.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
The issue MWeb.
image

**Comment by mananjadhav** [comment]
> The issue MWeb. 

I am not able to reproduce. Whats the filename and the device? I am trying on iPhone Safari.

Screenshot 2022-01-09 at 8 07 28 PM

Screenshot 2022-01-09 at 8 07 42 PM

**Comment by parasharrajat** [comment]
this firefox Android M web.

**Comment by mananjadhav** [comment]
Got it. Well, this issue exists with `ReportingTypingIndicator` as well currently.

Every time I've used this on the web, I also use `white-space: nowrap`, but wasn't sure if I should add right now. Adding it shows like this:

image

Do you have any other suggestions?

**Comment by parasharrajat** [comment]
Oh, Ok. If this issue exists on the main as well, I would report separately but now as we are working on a change that replaces both at once so we will have to fix it. We can't merge with a known bug. 

let's see if we can find a solution. But if you say that we should tackle this separately, then let me know. I am happy to report it.

**Comment by parasharrajat** [comment]
But I see why this is caused. and it's not your Code's fault. It's an issue with RN-web and I am reporting it there and There is another issue https://github.com/Expensify/App/issues/6913 due to that bug.

**Comment by parasharrajat** [comment]
Here is the RN-web issue link https://github.com/necolas/react-native-web/issues/2186

File: src/styles/styles.js [hunk 20]
```diff
@@ -2346,6 +2346,13 @@ const styles = {
         backgroundColor: colors.black,
         flex: 1,
     },
+
+    textWithEllipsis: {
+        textOverflow: 'ellipsis',
+        overflow: 'hidden',
+        ...whiteSpace.noWrap,
+        ...display.dInline,
```
Comment (parasharrajat) [inline src/styles/styles.js:2354]: It will break Native.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/styles/styles.js (hunk 20 above)
Comment (mananjadhav) [reply src/styles/styles.js:2354]: yeah I am aware. I am splitting it into two components.

Comment (mananjadhav) [reply src/styles/styles.js:2354]: Fixed

**Comment by mananjadhav** [comment]
@parasharrajat PR updated

Screenshot 2022-01-10 at 12 25 07 AM

> but now as we are working on a change that replaces both at once so we will have to fix it. We can't merge with a known bug.

Agreed and hence I've fixed it with a solution. Let me know if it works.

> you say that we should tackle this separately, then let me know. I am happy to report it.

While I've tackled the issue and if it is accepted, can we add a bonus to this job to solve the other bug as well?

**Comment by parasharrajat** [comment]
Ok, I have found the fix for the RN-web issue and I am going to submit a PR that will fix three issues at once in our repo including the one which you faced. 

let's wait for that PR to merge.

**Comment by mananjadhav** [comment]
@parasharrajat  Should I revert my recent changes? or wait for your PR to get merged? I saw your PR its the ideal fix for the issue.

**Comment by parasharrajat** [comment]
Le's wait for today. Thanks.

**Comment by mananjadhav** [comment]
@parasharrajat I can see the RNWeb PR is still not merged. Are you sure that this will fix the issue then I can probably revert my recent changes and close it from my side?

**Comment by parasharrajat** [comment]
@mananjadhav You can revert the changes for the ellipsis fix. PR is https://github.com/Expensify/App/pull/7254.

**Comment by mananjadhav** [comment]
@parasharrajat @Julesssss PR updated.

File: src/components/AttachmentModal.js [hunk 21]
```diff
@@ -73,6 +74,7 @@ class AttachmentModal extends PureComponent {
         this.submitAndClose = this.submitAndClose.bind(this);
         this.closeConfirmModal = this.closeConfirmModal.bind(this);
         this.isValidSize = this.isValidSize.bind(this);
+        this.splitExtensionFromFileName = this.splitExtensionFromFileName.bind(this);
```
Comment (parasharrajat) [inline src/components/AttachmentModal.js:77]: No extra bindings.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Review by parasharrajat — APPROVED** [review]
LGTM, tests well.

cc: @Julesssss 
:ribbon: :eyes: :ribbon:  C+ reviewed

**Review by Julesssss — APPROVED** [review]
Nice work. Looks good to me.

I was going to suggest a var name change, but it is too minor to care about for now.

**Pull request merged**